src/Indicator.vala

```diff
@@ -68,28 +64,43 @@ public class Notifications.Indicator : Wingpanel.Indicator {
             monitor.notification_received.connect (on_notification_received);
             monitor.notification_closed.connect (on_notification_closed);
 
-            dynamic_icon.button_press_event.connect ((e) => {
-                if (e.button == Gdk.BUTTON_MIDDLE) {
-                    notify_settings.set_boolean ("do-not-disturb", !notify_settings.get_boolean ("do-not-disturb"));
-                    return Gdk.EVENT_STOP;
-                }
-
-                return Gdk.EVENT_PROPAGATE;
-            });
-
             notify_settings.changed["do-not-disturb"].connect (() => {
                 set_display_icon_name ();
             });
-
-            nlist.add.connect (set_display_icon_name);
-            nlist.remove.connect (set_display_icon_name);
-
-            set_display_icon_name ();
         }
 
         return dynamic_icon;
     }
 
+    private async void get_session_notifications () throws ThreadError {
+        dynamic_icon.tooltip_markup = _("Updating session notifications.  Please wait");
+        dynamic_icon.get_style_context ().add_class ("disabled");
+
+        SourceFunc callback = get_session_notifications.callback;
+        ThreadFunc<bool> run = () => {
+            var previous_session = Session.get_instance ().get_session_notifications ();
+            previous_session.foreach ((notification) => {
+                nlist.add_entry (notification, false); // Do not need to write back to file
```

Comment:
You're adding a GTK widget from another thread here, this leads to crashes, you need to get the previous session in another thread but adding them needs to be done in the GTK thread

Reply:
Thanks for the review!  How much can be done in a separate thread? i.e. can widgets be created if they are not mapped? Is it emission of signals that causes the problem? It is filling the list box that causes most of the delay in the wingpanel appearing otherwise.   Perhaps we need a virtualizing list box??

Reply:
Note that the `nlist` is not realized or inserted into parent at the point it is updated in the thread and the indicator is disabled so it cannot be realized until the thread is finished.  I did not experience any crashes or error messages during testing.

Reply:
Now adding session notifications in Idle instead of thread.  Somewhat slower but still reasonably fast.